src/Features/Core/Portable/Completion/CompletionService.cs

```diff
@@ -138,7 +138,14 @@ public virtual TextSpan GetDefaultCompletionListSpan(SourceText text, int caretP
             string filterText)
         {
             var helper = CompletionHelper.GetHelper(document);
+            return FilterItems(helper, items, filterText);
+        }
 
+        internal static ImmutableArray<CompletionItem> FilterItems(
+            CompletionHelper completionHelper,
+            ImmutableArray<CompletionItem> items,
+            string filterText)
+        {
```

Comment:
why did we need to expose ths?  Why can't the person who needs this call into the method above?  It seems actually strange to expose this, because someone could call into this instead of hte virtual method which could be overridden.  That would be bypassing the intent here to allow the individual completion-service the ability to filter things.

Reply:
The method above requires a document. Now we have to support scenarios when we need to filter items we didn't create. Therefore, we do not have a document.

---
In reply to: [232508123](https://github.com/dotnet/roslyn/pull/29016#discussion_r232508123) [](ancestors = 232508123)

Reply:
That seems to violate one of the core designs behind roslyn language services (including completion). If it's C#/VB/F#/JS/TS, then we should have a Document.  If it's not one of those languages, then we shouldn't be interacting with teh code.

Reply:
Works for me.